core/controllers/topics_and_skills_dashboard.py

```diff
@@ -166,6 +166,73 @@ def get(self):
         self.render_json(self.values)
 
 
+class CategorizedAndUntriagedSkillsDataHandler(base.BaseHandler):
+    """Provides information about categorized skills and untriaged skill
+    summaries for the exploration editor page's skill editor component."""
+
+    GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
+    URL_PATH_ARGS_SCHEMAS = {}
+    HANDLER_ARGS_SCHEMAS = {
+        'GET': {}
+    }
+
+    @acl_decorators.open_access
+    def get(self):
```

Comment:
@seanlip I am a little bit confused here. The block of code in the new handler which I added is copied from the previous handler `TopicsAndSkillsDashboardPageDataHandler` which was being used earlier for fetching skills data for exploration editor page. I just deleted the unnecessary parts from the previous handler to make this new handler. Also, `skill_services.get_descriptions_of_skills()` already uses `get_multi`. So, do I need to optimise this?

Reply:
Quite possibly, because more people use the exploration editor than the topics and skills dashboard.

So my question still stands (with regards to the query complexity).

Reply:
@seanlip How about the following:

**For uncategorized skills**
```
        uncategorized_skills_id_and_topic_name = []
        for topic in topics:
            for skill_id in topic.uncategorized_skill_ids:
                uncategorized_skills_id_and_topic_name.append({
                    'skill_id': skill_id,
                    'topic_name': topic.name
                })
        uncategorized_skills_descriptions = (
            skill_services.get_descriptions_of_uncategorized_skills(
              uncategorized_skills_id_and_topic_name))
        uncategorized_skills_list = []
        for skill_data in uncategorized_skills_id_and_topic_name:
            skill_dict = {
                'skill_id': skill_data.get('skill_id'),
                'skill_description': (
                    uncategorized_skills_descriptions[
                        skill_data.get('skill_id')
                    ]
                ),
                'topic_name': skill_data.get('topic_name')
            }
            uncategorized_skills_list.append(skill_dict)
        categorized_skills_dict['uncategorized'] = uncategorized_skills_list
```

Reply:
**For uncategorized skills**
```
        categorized_skill_ids_with_topic_and_subtopic_name = []
        for topic in topics:
            subtopics = topic.subtopics
            for subtopic in subtopics:
                for skill_id in subtopic.skill_ids:
                    categorized_skill_ids_with_topic_and_subtopic_name.append({
                        'skill_id': skill_id,
                        'topic_name': topic.name,
                        'subtopic_title': subtopic.title
                    })
        categorized_skill_descriptions = (
            skill_services.get_descriptions_of_categorized_skills(
                categorized_skill_ids_with_topic_and_subtopic_name))
        categorized_skills_list = []
        for skill_data in categorized_skill_ids_with_topic_and_subtopic_name:
            skill_dict = {
                'skill_id': skill_data.get('skill_id'),
                'skill_description': (
                    categorized_skill_descriptions[
                        skill_data.get('skill_id')
                    ]
                ),
                'topic_name': skill_data.get('topic_name'),
                'subtopic_title': skill_data.get('subtopic_title')
            }
            categorized_skills_list.append(skill_dict)
        categorized_skills_dict['categorized'] = categorized_skills_list
```

Reply:
By doing this, we can move the function which interacts with the datastore outside the for-loop and complete all the datastore functionality with one `get-multi` call (previously, for each topic and subtopic, datastore calls were made and now, all the calls have been shifted to one). Please let me know if more optimizations can be done on this.

Reply:
With a little change in the frontend, we can use this format.

Reply:
I scanned through this briefly and I think the approach seems reasonable to me. If you can get it down to 2 get-multi calls, that shouldn't be an issue.

The only suggestion I have is to try and avoid duplicating code, where possible. It will make the resulting logic easier to maintain.

Thanks for digging into this!

Reply:
Also, just a reminder -- if you need me to look at something, please assign me, otherwise it might slip. (Ditto for anyone else whom you need to take an action on the PR.) Thanks!